### ISSUE TITLE
More complete Meteor.userId implementation on the server side

### ISSUE THREAD

**Issue by cwohlman** [meteor/meteor#5697]
Currently accessing `Meteor.userId` on the server is somewhat of a hit and miss operation. There are at least two more uses cases for `Meteor.userId` which I think would be a good idea to support:

1. In publish methods.
2. In setTimeout/setImmediate/setInterval calls within Meteor methods (see https://github.com/meteor/meteor/commit/1a503d1abf42abf39e7dd056f9f6497a3d15d3c4 for why this doesn't happen).

I'm thinking it should be pretty simple to implement:

1. Create a `Meteor.EnvironmentVariable` to hold the current userId, e.g. `Meteor._userId = new Meteor.EnvironmentVariable`
2. Wrap method & subscription handlers in `withValue`, e.g. `Meteor._userId.withValue(currentUserId, handler)`.
3. Since the userId can change while a method handler is being run, perhaps we should store the whole object referencing the userId inside of the EnvironmentVariable, e.g. `Meteor._userIdReference.withValue(DDP._currentInvocation.get())`, this would preserve the userId when someone uses a timer callback , but would allow the userId to change when the userId from the currentInvocation changes (i.e. would mimic current behavior). For publish functions we can simply store an object with the current userId, e.g. `Meteor._userIdReference.withValue({ userId: this.userId })`, obviously the code for Meteor.userId would then be pretty simple: `return Meteor._userIdReference.get().userId`

I'm interested in writing this code myself, but I have a few questions:

1. Are there any use cases that depend on `Meteor.userId()` throwing an error when called outside of a method invocation? In other words, would this be considered a backwards-incompatible change?
2. Is this something that people are interested in?

**Comment by fabiodr** [comment]
Yes, because of this I can't reuse some model logic that is called from diffent contexts, like from a template helper and a Meteor method. I had to send the this.userId as argument when inside a publication or method. Is there some workaround or better solution?

**Comment by stubailo** [comment]
> I had to send the this.userId as argument when inside a publication or method.

I would argue that we should actually move in the other direction, and encourage people to pass arguments more explicitly rather than relying on global context. This sort of global context variable often causes problems in asynchronous code on the server, as well as with server-side rendering, etc; it also makes all of the Meteor internal code more complex.

So I think such a pull request wouldn't be a good idea at the moment for two reasons:

1. It's not clear if it's a good idea to make Meteor.userId more powerful - I think this should be opened to a wider discussion.
2. It would require working in Meteor internals and could potentially result in security bugs, since probably many people would be using this for security checks against the logged in user. This means we should involve someone who has a lot more experience with these systems in the conversation. But only after we resolve (1).

**Comment by fabiodr** [comment]
Most of the times i needed userId had relation to store data in the scope of: user, user logged session, user connection (different tabs), only a ddp request and some combination of them. There are some packages for handling this specifics lifespans but they aren't solid. The guide will provide some information about it?

**Comment by stubailo** [comment]
I'm not exactly sure what you mean - could you provide some specific use
cases?

**Comment by mitar** [comment]
I made a package for this: https://github.com/peerlibrary/meteor-user-extra

**Comment by mitar** [comment]
Also, this is probably a duplicate of #5697.

**Comment by stubailo** [comment]
Duplicate of #5462
